libsolidity/codegen/ir/IRGenerator.cpp

```diff
@@ -382,34 +382,53 @@ string IRGenerator::generateGetter(VariableDeclaration const& _varDecl)
 			{
 				if (returnTypes[i]->category() == Type::Category::Mapping)
 					continue;
-				if (auto arrayType = dynamic_cast<ArrayType const*>(returnTypes[i]))
+				auto const* arrayType = dynamic_cast<ArrayType const*>(returnTypes[i]);
+				if (arrayType)
 					if (!arrayType->isByteArray())
 						continue;
 
-				// TODO conversion from storage byte arrays is not yet implemented.
 				pair<u256, unsigned> const& offsets = structType->storageOffsetsOfMember(names[i]);
 				vector<string> retVars = IRVariable("ret_" + to_string(returnVariables.size()), *returnTypes[i]).stackSlots();
 				returnVariables += retVars;
 				code += Whiskers(R"(
 					<ret> := <readStorage>(add(slot, <slotOffset>))
 				)")
 				("ret", joinHumanReadable(retVars))
-				("readStorage", m_utils.readFromStorage(*returnTypes[i], offsets.second, true))
+				(
+					"readStorage",
+					arrayType ?
+						m_utils.copyArrayFromStorageToMemoryFunction(
+							dynamic_cast<ArrayType const&>(*arrayType->copyForLocation(DataLocation::Storage, false)),
+							*arrayType
+						) :
+						m_utils.readFromStorage(*returnTypes[i], offsets.second, true)
+				)
 				("slotOffset", offsets.first.str())
 				.render();
 			}
 		}
 		else
 		{
 			solAssert(returnTypes.size() == 1, "");
+			auto const* arrayType = dynamic_cast<ArrayType const*>(returnTypes.front());
+			if (arrayType)
+				solAssert(arrayType->isByteArray(), "");
 			vector<string> retVars = IRVariable("ret", *returnTypes.front()).stackSlots();
 			returnVariables += retVars;
-			// TODO conversion from storage byte arrays is not yet implemented.
 			code += Whiskers(R"(
```

Comment:
Maybe it would be better to use different templates for arrays and non-arrays?

Reply:
Or maybe calling it from `readFromStorageReferenceType`, so we do not add anything here.
That way it activates more tests as well, and implements more features we are missing

Reply:
Sounds good!